src/zino/planned_maintenance.py

```diff
@@ -0,0 +1,217 @@
+        task until `now`
+        """
+        if self.last_run:
+            return [
+                pm for pm in self.planned_maintenances.values() if self.last_run < pm.start_time <= now < pm.end_time
+            ]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.start_time <= now < pm.end_time]
+
+    def get_ended_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that have ended since the last run of this
+        task until `now`
+        """
+        if self.last_run:
+            return [pm for pm in self.planned_maintenances.values() if self.last_run < pm.end_time <= now]
+        else:
+            return [pm for pm in self.planned_maintenances.values() if pm.end_time <= now]
+
+    def get_active_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that are currently active
+
+        This means it has started before `now` and the end_time is later than `now`
+        """
+        return [pm for pm in self.planned_maintenances.values() if pm.start_time < now < pm.end_time]
+
+    def get_old_planned_maintenances(self, now: datetime.datetime) -> list[PlannedMaintenance]:
+        """Returns all planned maintenances that should get deleted
+
+        This means that `now` is 3 days later than end_time
+        """
+        return [pm for pm in self.planned_maintenances.values() if now - pm.end_time > datetime.timedelta(days=3)]
+
+    def add_pm_observer(self, observer: PlannedMaintenanceObserver) -> None:
+        """Adds an observer function that will be called any time a planned maintenance
+        is added or removed
+        """
+        self._observers.append(observer)
+
+    def get_next_available_pm_id(self) -> int:
+        """Returns the next available planned maintenance id"""
+        self.last_pm_id += 1
+        return self.last_pm_id
+
+    def _call_observers(self) -> None:
+        for observer in self._observers:
+            observer()
+
+    def periodic(self, state: "ZinoState"):
+        now = datetime.datetime.now()
+
+        # Initiate PM once it becomes active
+        for started_pm in self.get_started_planned_maintenances(now=now):
+            self._start(state, started_pm)
+
+        # Make sure all events that matches a PM is ignored
+        for event in state.events.events.values():
+            self._check_event(state, event, now)
+
+        # Set events matching ended PMs to open
+        for ended_pm in self.get_ended_planned_maintenances(now=now):
+            self._end(state, ended_pm)
+
+        old_pms = self.get_old_planned_maintenances(now)
+        for pm in old_pms:
+            self.close_planned_maintenance(pm.id, "timer expiry for old PMs", "zino")
+
+        self.last_run = now
+
+    def _start(self, state: "ZinoState", pm: PlannedMaintenance):
```

Comment:
Start what?

Reply:
this is moved into Model class. The name remains but makes more sense in the context of being used on an PM object